crates/bevy_ecs/src/system/combinator.rs

```diff
@@ -182,18 +180,16 @@ where
         )
     }
 
-    fn run<'w>(&mut self, input: Self::In, world: &'w mut World) -> Self::Out {
-        // SAFETY: Converting `&mut T` -> `&UnsafeCell<T>`
-        // is explicitly allowed in the docs for `UnsafeCell`.
-        let world: &'w UnsafeCell<World> = unsafe { std::mem::transmute(world) };
+    fn run(&mut self, input: Self::In, world: &mut World) -> Self::Out {
+        let world = world.as_unsafe_world_cell();
         Func::combine(
             input,
             // SAFETY: Since these closures are `!Send + !Sync + !'static`, they can never
             // be called in parallel. Since mutable access to `world` only exists within
             // the scope of either closure, we can be sure they will never alias one another.
```

Comment:
Slight note on this comment: it is not true that `!Send + !Sync + !'static` prevents calling them at the same time, since it doesn't rule out calling them reentrantly. This is pretty hard to achieve with `!'static` closures that you cannot control, but in theory it's not impossible. For example `Func` might store one in a scoped thread local and run the other one, which then internally gets the one stored in the thread local and run it reentrantly, ultimately creating two mutable borrows of the world at the same time.

Reply:
I've opened #14709 separately for this issue, this can be ignored for this PR.